src/FastRouteRouter.php

```diff
@@ -277,41 +255,80 @@ public function generateUri($name, array $substitutions = [], array $options = [
         }
 
-        foreach ($substitutions as $key => $value) {
-            $pattern = sprintf(
-                '~%s~x',
-                sprintf(self::VARIABLE_REGEX, preg_quote($key))
-            );
-            $path = preg_replace($pattern, $value, $path);
-        }
+        $routeParser        = new RouteParser();
+        $routes             = array_reverse($routeParser->parse($route->getPath()));
+        $requiredParameters = [];
+
+        // One route pattern can correspond to multiple routes if it has optional parts
+        foreach ($routes as $parts) {
+            // Check if all parameters can be substituted
+            $requiredParameters = $this->hasRequiredParameters($parts, $substitutions);
+
+            // If not all parameters can be substituted, try the next route
+            if ($requiredParameters !== true) {
+                continue;
+            }
+
+            // Generate the path
+            $path = '';
+            foreach ($parts as $part) {
+                if (is_string($part)) {
+                    // Append the string
+                    $path .= $part;
+                    continue;
+                }
 
-        // 1. remove optional segments' ending delimiters
-        //    and remove leftover regex char classes like `:[` and `:prefix-[` (issue #18)
-        // 2. split path into an array of optional segments and remove those
-        //    containing unsubstituted parameters starting from the last segment
-        $path = preg_replace('/\]|:.*\[/', '', $path);
-        $segs = array_reverse(explode('[', $path));
-        foreach ($segs as $n => $seg) {
-            if (strpos($seg, '{') !== false) {
-                if (isset($segs[$n - 1])) {
-                    throw new Exception\InvalidArgumentException(
-                        'Optional segments with unsubstituted parameters cannot '
-                        . 'contain segments with substituted parameters when using FastRoute'
-                    );
+                // Check substitute value with regex
+                $regex = '~^' . str_replace('/', '\/', $part[1]) . '$~';
+                if (! preg_match($regex, $substitutions[$part[0]])) {
+                    throw new Exception\InvalidArgumentException(sprintf(
+                        'Parameter value for [%s] did not match the regex `%s`',
+                        $part[0],
+                        $part[1]
+                    ));
                 }
-                unset($segs[$n]);
+
+                // Append the substituted value
+                $path .= $substitutions[$part[0]];
+            }
+
+            // Return generated path
+            return $path;
+        }
+
+        // No valid route was found; explain which minimal required parameters are needed
+        throw new Exception\InvalidArgumentException(sprintf(
+            'Expected parameter values for at least [%s], but received [%s]',
+            implode(',', $requiredParameters),
```

Comment:
Returning all possible combinations means that it always has to go through all possible combinations for that route. Plus is might also get confusing when return all combinations. Besides that, it's up to the developer to figure it out for complex routes. But what I can do is adding the route name to the exception so he can easily track what route is throwing the error.